src/main/java/org/jitsi/videobridge/BitrateController.java

```diff
@@ -0,0 +1,568 @@
+            }
+
+            return pkts;
+        }
+    }
+
+    /**
+     * The {@link PacketTransformer} that handles incoming/outgoing RTCP
+     * packets for this {@link BitrateController} instance. Internally,
+     * it delegates this responsibility to the appropriate media stream track
+     * bitrate controller ({@link SimulcastController}, etc).
+     */
+    class RTCPTransformer
+        implements PacketTransformer
+    {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void close()
+        {
+
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public RawPacket[] reverseTransform(RawPacket[] pkts)
+        {
+            return pkts;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public RawPacket[] transform(RawPacket[] pkts)
+        {
+            return pkts;
+        }
+    }
+
+    /**
+     * Computes the optimal and the target bitrate, limiting the target to be
+     * less than bandwidth estimation specified as an argument.
+     *
+     * @param bweBps the max bandwidth estimation that the target bitrate must
+     * not exceed.
+     *
+     * @return an array of {@link EndpointBitrateAllocation}.
+     */
+    private EndpointBitrateAllocation[] allocate(
+        long bweBps, List<Endpoint> conferenceEndpoints)
+    {
+        // Init.
+        int szConference = conferenceEndpoints.size();
+
+        // subtract 1 for destEndpoint.
+        EndpointBitrateAllocation[] endpointBitrateAllocations
+            = new EndpointBitrateAllocation[szConference - 1];
+
+        int lastN = dest.getLastN();
+        if (lastN < 0)
+        {
+            // If lastN is disable, pretend lastN == szConference.
+            lastN = endpointBitrateAllocations.length;
+        }
+
+        Endpoint destEndpoint = dest.getEndpoint();
+        Set<String> pinnedEndpoints = destEndpoint.getPinnedEndpoints();
+        int szPinned = pinnedEndpoints == null ? 0 : pinnedEndpoints.size();
+        if (szPinned > 0)
+        {
+            // The lastN value should be large enough to accommodate all the
+            // pinned endpoints.
+            lastN = Math.max(lastN, szPinned);
+        }
+
+        Set<String> selectedEndpoints = destEndpoint.getSelectedEndpoints();
+        int offPinned = 0, szSelected
+            = selectedEndpoints == null ? 0 : selectedEndpoints.size();
+
+        int maxQuality = 0;
+        long oldBweBps = bweBps;
+
+        int idx = -1;
+        for (int i = 0; i < szConference; i++)
+        {
+            Endpoint sourceEndpoint = conferenceEndpoints.get(i);
+            if (sourceEndpoint == destEndpoint)
+            {
+                continue;
+            }
+
+            idx++;
+
+            boolean pinned = szPinned != 0
+                && pinnedEndpoints.contains(sourceEndpoint.getID());
+
+            int priority; // Higher priority endpoints are optimized first.
```

Comment:
I wanted to avoid all these sorting operations and do everything in a single loop, but I'm happy to sacrifice a little performance for improved readability.